juju/juju
# Add display name to instance data

This commit makes it possible for provider code to pass through a display name when instances are created.

Implementations of `InstanceBroker` can add `DisplayName` `string` to their `StartInstanceResult`
instance that's returned with `StartInstance`. That string is propagated through to stat as part of `instanceData` and surfaced up when end-users call `juju status`.

The PR touches many files as it has required some reworking of a few of the provisioning mechanics through each stage of the API. Here are the major public API changes:

* The [`Machine.SetProvisioned`](https://github.com/timClicks/juju/blob/f4b75139fcfef12af7a0b8a6e13d96b216de92fe/state/machine.go#L1436) method takes a new parameter, `displayName string`
* the `MachineProvisioner` interface has been modified. [`MachineProvisioner.SetInstanceInfo`](https://github.com/timClicks/juju/blob/f4b75139fcfef12af7a0b8a6e13d96b216de92fe/api/provisioner/machine.go#L86) now includes a `displayName string` argument. 

**Backwards compatibility**

_Provider code_

No changes are required to code that does not wish to make use of instance display names. The default behaviour is to revert to the instance ID. Current provider code remains unchanged.

_API internals_

Adding a new field may require bumping the façade number. This hasn't yet been done within this PR. 

**Future work**

* Add the ability to create a custom display name via the CLI `juju set-display-name`
* Extend the functionality into `state/addmachine.go`. `addmachine.go` does not alter `instanceData`. Provisioners that make use of this code to provision machines—generally the manual provisioners—won't be able to set an instance's display name without more work.  
 
## QA steps

Bootstrap into Oracle provider, add a machine and see that the `Inst id` column is only 6 significant "figures":

```plain
$ juju bootstrap --config compartment-id=ocid1... oracle tsm-phx
...

$ juju add-machine -n 2
created machine 0
created machine 1

$ juju status
Model    Controller  Cloud/Region         Version      SLA          Timestamp
default  tsm-phx     oracle/us-phoenix-1  2.6-beta1.1  unsupported  21:07:57+13:00

Machine  State    DNS              Inst id    Series  AZ  Message
0        pending  129.146.175.166  ...pbcehq  bionic      running
1        pending  129.146.173.212  ...6qhlsq  bionic      running

$ juju machines --format=yaml
model: default
machines:
  "0":
    juju-status:
      current: started
      since: 10 Jan 2019 21:10:07+13:00
      version: 2.6-beta1.1
    dns-name: 129.146.175.166
    ip-addresses:
    - 129.146.175.166
    - 10.0.1.2
    instance-id: ocid1.instance.oc1.phx.abyhqljtnhgn33m35rurkociav3aoyrgdqrwgfngoungqqjwzgoy6mpbcehq
    display-name: '...pbcehq'
    machine-status:
      current: running
      message: running
      since: 10 Jan 2019 21:07:52+13:00
    series: bionic
    network-interfaces:
      ens3:
        ip-addresses:
        - 10.0.1.2
        mac-address: 00:00:17:02:cb:3a
        gateway: 10.0.1.1
        is-up: true
    hardware: arch=amd64
...
```


## Documentation changes

The output for the MAAS and OCI providers will look different _when interacting with newly provisioned machines_. Machines that have already been provisioned will still have the same display name.

This might create an awkward situation within `juju show-status` where some instances will have long, ugly instance ids, whereas others have a shorter one.

## Bug reference

* https://bugs.launchpad.net/juju/+bug/1792210
